crates/turborepo-lib/src/package_graph/mod.rs

```diff
@@ -63,8 +64,44 @@ impl PackageGraph {
         PackageGraphBuilder::new(repo_root, root_package_json)
     }
 
-    pub fn validate(&self) -> Result<()> {
-        // TODO
+    pub fn validate(&self) -> Result<(), builder::Error> {
+        // This is equivalent to AcyclicGraph.Cycles from Go's dag library
+        let cycles = petgraph::algo::tarjan_scc(&self.workspace_graph)
+            .into_iter()
+            .filter(|cycle| cycle.len() > 1)
+            .collect::<Vec<_>>();
```

Comment:
We can probably just map here and use itertools::join? Same cost for the golden path, and one fewer alloc for the error case.

https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.join

```rust
let cycles_output = petgraph::algo::tarjan_scc(&self.workspace_graph)
            .into_iter()
            .filter(|cycle| cycle.len() > 1)
            .map(|cycle| {
                // format string
            })
            .join("\n");

if !cycles.is_empty() { // checking against a string instead
    return Err(builder::Error::CyclicDependencies(cycles_output))
}
```